### PR TITLE
Added preconditioning functionality

### PR DESCRIPTION
**Pull request by sfetzel**
I have the following use case:

I want to use the Hessian as a preconditioner only, therefore replacing the gradient $g$ with $H^{-1}\cdot g$. After preconditioning, any optimizer can apply the gradients to the parameters, e.g. Adam or SGD.

Therefore I added code, which replaces the gradients with $H^{-1}\cdot g$ (essentially the result of CG), when the learning rate is zero.

### CODE REVIEW

**Comment by ltatzel** [comment]
Hi Simon, 

Sorry for the late reply and thanks a lot for the PR. I think, however, that your use case is quite niche. Also, I'm a bit worried about replacing the parameters' `grad` field with the Newton step (that might cause problems in other use cases that assume `grad` to contain the gradient) and the way this feature is "activated" (via `lr=0` in the optimizer's constructor). 

In your use case, I would suggest something like:
```python
import torch
from example_utils import get_small_nn_testproblem
from torch.nn.utils import parameters_to_vector

from hessianfree.optimizer import HessianFree
from hessianfree.utils import vector_to_parameter_list

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")


if __name__ == "__main__":

    print(f"\nRunning example on DEVICE = {DEVICE}")

    torch.manual_seed(0)

    # Set up problem
    model, (inputs, targets), loss_function = get_small_nn_testproblem(device=DEVICE)

    # Set up optimizer
    opt = HessianFree(
        model.parameters(),
        lr=1.0,  # <-- use `1.0` for Newton step
        use_linesearch=False,
        adapt_damping=False,
        verbose=True,
    )

    def forward():
        outputs = model(inputs)
        loss = loss_function(outputs, targets)
        return loss, outputs

    # Compute Newton step
    params_old = parameters_to_vector(model.parameters())
    opt.step(forward=forward)
    params_new = parameters_to_vector(model.parameters())
    newton_step = vector_to_parameter_list(params_new - params_old, model.parameters())

    # Replace `grad` field with Newton step
    for param, newton_step_param in zip(model.parameters(), newton_step):
        param.grad = newton_step_param

    # Updated `grad` field
    print("\nUpdated `grad` field:")
    for param in model.parameters():
        print(param.grad)
```

I hope that helps. Let me know if you have further questions or suggestions.

Best, 
Lukas
